crates/router/src/core/payment_methods/cards.rs

```diff
@@ -896,13 +994,248 @@ pub async fn get_client_secret_or_add_payment_method(
+    req: api::PaymentMethodIntentCreate,
+    merchant_account: &domain::MerchantAccount,
+    key_store: &domain::MerchantKeyStore,
+) -> errors::RouterResponse<api::PaymentMethodResponse> {
+    let db = &*state.store;
+    let merchant_id = merchant_account.get_id();
+    let customer_id = req.customer_id.to_owned();
+
+    db.find_customer_by_merchant_reference_id_merchant_id(
+        &(state.into()),
+        &customer_id,
+        merchant_account.get_id(),
+        &key_store,
+        merchant_account.storage_scheme,
+    )
+    .await
+    .to_not_found_response(errors::ApiErrorResponse::CustomerNotFound)?;
+
+    let payment_method_billing_address: Option<Encryptable<Secret<serde_json::Value>>> = req
+        .billing
+        .clone()
+        .async_map(|billing| create_encrypted_data(state, key_store, billing))
+        .await
+        .transpose()
+        .change_context(errors::ApiErrorResponse::InternalServerError)
+        .attach_printable("Unable to encrypt Payment method billing address")?;
+
+    // create pm entry
+
+    // Need a global_id type for this
+    let payment_method_id = generate_id(consts::ID_LENGTH, "pm");
+    let payment_method = create_payment_method_for_intent(
+        state,
+        &req,
+        &customer_id,
+        payment_method_id.as_str(),
+        merchant_id,
+        key_store,
+        merchant_account.storage_scheme,
+        payment_method_billing_address.map(Into::into),
+    )
+    .await
+    .attach_printable("Failed to add Payment method to DB")?;
+
+    let resp = payment_methods::generate_payment_method_response(&payment_method)?;
+
+    Ok(services::ApplicationResponse::Json(resp))
+}
+
+#[cfg(all(feature = "v2", feature = "payment_methods_v2"))]
+#[instrument(skip_all)]
+pub async fn payment_method_intent_confirm(
+    state: &routes::SessionState,
+    req: api::PaymentMethodIntentConfirm,
+    merchant_account: &domain::MerchantAccount,
+    key_store: &domain::MerchantKeyStore,
+    pm_id: String,
+) -> errors::RouterResponse<api::PaymentMethodResponse> {
+    req.validate()?;
+
+    let db = &*state.store;
+    let client_secret = req.client_secret.clone();
+
+    let payment_method = db
+        .find_payment_method(
+            &(state.into()),
+            &key_store,
+            &pm_id,
+            merchant_account.storage_scheme,
+        )
+        .await
+        .change_context(errors::ApiErrorResponse::PaymentMethodNotFound)
+        .attach_printable("Unable to find payment method")?;
+
+    let customer_id = payment_method.customer_id.to_owned();
+
+    db.find_customer_by_merchant_reference_id_merchant_id(
+        &(state.into()),
+        &customer_id,
+        merchant_account.get_id(),
+        &key_store,
+        merchant_account.storage_scheme,
+    )
+    .await
+    .to_not_found_response(errors::ApiErrorResponse::CustomerNotFound)?;
+
+    if authenticate_pm_client_secret_and_check_expiry(&client_secret, &payment_method)? {
```

Comment:
Move this check before find customer and use when function instead of if return.